%\VignetteIndexEntry{Trendy Vignette}
%\VignettePackage{Trendy}
%\VignetteEngine{knitr::knitr}

\documentclass{article}
\usepackage{graphicx, graphics, epsfig,setspace,amsmath, amsthm}
\usepackage{natbib}
\usepackage{moreverb}
\usepackage{float}

<<style-knitr, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@

\begin{document}

\title{Trendy: Segmented regression analysis of
expression dynamics for high-throughput ordered
profiling experimentsa}
\author{Rhonda Bacher, Ning Leng, Ron Stewart}
\maketitle
\tableofcontents
\setcounter{tocdepth}{2}



\section{Overview}
\label{sec:intro}
Trendy is an R package that can be used to perform breakpoint analysis on microarray or RNA-seq expression data with ordered conditions (e.g. time-course, spatial-course).
For each gene or other features, Trendy estimates the optimal number of breakpoints as well as
the breakpoints by fitting a set of segmented regression models.
The top dynamic genes are then identified by taking genes that can be well profiled by its gene-specific segmented regression model. Trendy also implements functions to visualize the dynamic genes and their trends, to order 
dynamic genes by their trends, and to compute breakpoint distribution at different
time-points (e.g. detect time-points with a large number of expression changes).

\subsection{The model}
To illustrate Trendy, here we use time course gene expression data as an example. 
Although, Trendy may also be applied to other types of features (e.g. isoform or exon expression) and/or other
types of experiments with ordered conditions (e.g. spatial course).

Denote the normalized gene expression of gene $g$ and sample/time $t$ as $Y_{g,t}$.
Denote the total number of genes as $G$ and the total number of samples/times as $N$.
For each gene, Trendy fits segmented regression models with varying numbers of breakpoints from 1 to $K$. $K$ defaults to 3 but can also
be specified by the user. The segmented R package is used to fit the segmented regression models.

For a given gene, among the models with varying $k$, Trendy selects the optimal
number of breakpoints for this gene by comparing the coefficient of determination ($R^2$) for each model.

To avoid overfitting, the optimal number of breakpoints will be set as
$\tilde{k_g} = \tilde{k_g} -1$ if any of the following happens:
at least one segment has less than $c_{num}$ samples
or $R^{2}_{g,\tilde{k_g}} - R^{2}_{g,\tilde{k_g}-1} < c_{diff}$. The thresholds
$c_{num}$ and $c_{diff}$ can be specified by the user; the 
defaults are 5 and 0.1,
respectively.


Trendy reports the following for the optimal model:
\begin{itemize}
  \item Gene specific adjusted $R^2$ (penalized for the chosen value of $k$)
  \item Segment slopes
  \item Breakpoint estimates
\end{itemize}

Among all genes, the top dynamic genes are defined as those whose optimal model has high adjusted $R^2$s.

To compute the breakpoint distribution over the time-course, Trendy calculates the number of 
breakpoints for each time-point across all the genes.

The time-points with high $D_t$ can be considered as those with global expression changes.

Trendy also summarizes the fitted trend or expression pattern of top genes. For samples between the $i^{th}$
and $i+1 ^{th}$ breakpoint for a given gene, if the t-statistic of
the segment slope has p-value greater than $c_{pval}$,
the trend of this segment will be defined as no change. Otherwise the
trend of this segment will be defined as up/down based on the slope coefficient.
The default value of $c_{pval}$ is 0.1, but may also be specified by the user.

\section{Installation}

\subsection{Install via GitHub}

The Trendy package can be installed using functions in the devtools package.

To install, type the following code into R:
<<eval=FALSE, echo=TRUE, cache=FALSE, message=FALSE, warning=FALSE>>=
install.packages("devtools")
library(devtools)
install_github("rhondabacher/Trendy/package/Trendy")
@

\subsection{Install locally}

Install packages segmented and gplots:


<<eval=FALSE, echo=TRUE>>=
install.packages(c("segmented","gplots"))
library("segmented")
library("gplots")
@

Download the Trendy package from: \url{https://github.com/rhonda/Trendy/tree/master/package}

And install the package locally.


\subsection{Load the package}

To load the Trendy package:
<<eval=TRUE, echo=TRUE, message=FALSE>>=
library(Trendy)
@

\section{Analysis}

\subsection{Input}

The input data should be a $G-by-N$ matrix containing the expression values for each gene and
 each sample, where $G$ is the number of genes and $N$ is the number of samples. 
The samples should be sorted following the time course order.
These values should exhibit expression data after normalization across samples. 
For example, for RNA-seq data, the raw counts may be normalized using MedianNorm and
\Rfunction{GetNormalizedMat} function in \Biocpkg{EBSeq}. More details can be found in the \Biocpkg{EBSeq} vignette:

\url{http://www.bioconductor.org/packages/devel/bioc/vignettes/EBSeq/inst/doc/EBSeq_Vignette.pdf}


The object TrendyExData is a simulated data matrix containing 50 rows of genes and 40 columns of samples.

<<eval=TRUE>>=
data(TrendyExData)
str(TrendyExData)
@




\subsection{Run Trendy}

The \Rfunction{trendy} function will fit multiple segmented regressions model for each gene (via the \CRANpkg{segmented} R package) and select the the optimal model. Here we want to only consider a maximum of two breakpoints for each gene.

<<eval=TRUE>>=
res <- trendy(TrendyExData, Max.K = 2)
res.top <- toptrendy(res) 
# default adjusted R square cutoff is 0.5
res.top$radj
@

The \Rfunction{toptrendy} function may be used to extract top dynamic genes.
By default, \Rfunction{toptrendy} will extract genes whose adjusted $R^2$, $\bar{R}^{2}$,
is greater or equal to 0.5. To change this threshold, a user may specify the
\Rcode{AdjR.Cut} parameter in the \Rfunction{toptrendy} function.
res.top\$radj gives the $\bar{R}^{2}$ of the top dynamic genes sorted decreasingly.


By default the \Rfunction{trendy} function only considers genes whose mean expression is greater than 10. 
To use another threshold, the user may specify the parameter \Robject{Mean.Cut}.


\subsection{Visualize trends of the top dynamic genes}

res.top\$id.sign gives the trend specification of the top genes. The function \Rfunction{trendheatmap} can be used to display these trends:

<<eval=TRUE, warning=FALSE, fig.width=7, fig.align='center', out.width='.6\\textwidth'>>=
res.trend <- trendheatmap(res.top)
str(res.trend)
@

The \Rfunction{trendheatmap} function classifies the top dynamic genes into
three groups: start with up, start with down and start with no change.
Within each group, genes are sorted by the position of the first breakpoint.


To generate an expression heatmap of the first group of genes (first go up):

<<eval=TRUE, warning=FALSE, fig.width=7, fig.align='center', out.width='.8\\textwidth', message=FALSE>>=
library(gplots)
heatmap.2(TrendyExData[names(res.trend$firstup),],
  trace="none", Rowv=F,Colv=F,
	scale="row", main="top genes (first go up)")
@


Similarly, to generate an expression heatmap of the second group of genes (first go down):
<<eval=TRUE, warning=FALSE, fig.width=7, fig.align='center', out.width='.8\\textwidth'>>=
heatmap.2(TrendyExData[names(res.trend$firstdown),],
  trace="none", Rowv=F,Colv=F,
	scale="row", main="top genes (first go down)")
@

To generate an expression heatmap of the second group of genes (first no change):
<<eval=TRUE, warning=FALSE, fig.width=7, fig.align='center', out.width='.8\\textwidth'>>=
heatmap.2(TrendyExData[names(res.trend$firstnochange),],
  trace="none", Rowv=F,Colv=F,
	scale="row", main="top genes (first no change)",
	cexRow=.8)
@


\subsection{Visualize individual genes}
The \Rfunction{plotfeature} function may be used to plot expression of individual features/genes and the fitted lines.

For example, to plot the top six genes in the first group of genes (first go up):

<<eval=TRUE, warning=FALSE, fig.width=7, fig.align='center', out.width='1\\textwidth'>>=
par(mfrow=c(3,2))
plot1 <- plotfeature(TrendyExData, 
                    Feature.Names = names(res.trend$firstup)[1:6], 
                    Seg.Data = res)
@

The input of function \Rfunction{plotfeature} requires the expression data and a list of genes of interest.
The parameter \Robject{Seg.Data} results from the \Rfunction{trendy} function.
If it is not specified, then \Rfunction{plotfeature} will run \Rfunction{trendy} on the genes of interest before plotting.
Specifying the fitted results obtained from previous steps will save time by avoiding fitting the models again.

Similarly, to plot the top six genes in the second group of genes (first go down):

<<eval=TRUE, warning=FALSE, fig.width=7, fig.align='center', out.width='1\\textwidth'>>=
par(mfrow=c(3,2))
plot2 <- plotfeature(TrendyExData, 
                    Feature.Names = names(res.trend$firstdown)[1:6],
                    Seg.Data = res)
@

To plot the two genes in the third group of genes (first no change):
<<eval=TRUE, warning=FALSE, fig.height=5, fig.width=10, fig.align='center'>>=
par(mfrow=c(1,2))
plot2 <- plotfeature(TrendyExData, 
                    Feature.Names = names(res.trend$firstnochange)[1:2],
                    Seg.Data = res)
@

\subsection{Gene specific estimates}

For a given gene of interest, its estimated parameters can be obtained by (using g2 as an example):
<<eval=TRUE, fig.height=3, fig.width=4, fig.align='center'>>=
par(mfrow=c(1,1))
plot2 <- plotfeature(TrendyExData, 
                    Feature.Names = "g2",
                    Seg.Data = res)
print(res.top$bp["g2"]) # break points
print(res.top$radj["g2"]) # adjusted r squared
print(res.top$slp["g2"]) # fitted slopes of the segments
print(res.top$slp.pval["g2"]) # p value of each the segment
@

The above printouts show that for gene g2 the optimal number of breakpoints
is 2. Two estimated breakpoints are around time-points s12 and s30.
The fitted slopes for the 3 segments are 3.31, 0.06 and -2.97, which indicate the trend is up-same-down.


These estimates can also be automatically formatted using the function \Rfunction{formatresults} 
which can be saved as a .txt. or .csv file. The output currently includes the estimate slope of 
each segment, the estimated breakpoint, and the adjusted $R^2$.

<<eval=TRUE>>=
trendy.summary <- formatresults(res.top)
head(trendy.summary)

# write.table(trendy.summary, file="trendy_summary.txt")
@
The NA indicates that g3 does not have a slope3 since it only has one breakpoint (i.e two segments).


\subsection{Breakpoint distribution over the time course}

To calculate number of breakpoints over the time course:

<<eval=TRUE, warning=FALSE, fig.height=3.5, fig.width=7, fig.align='center', out.width='.8\\textwidth'>>=
res.bp <- bpdist(res.top)
barplot(res.bp, ylab="Number of breakpoint", col="blue")
@

The bar plot indicates that a number of genes have breakpoints around s12 and s13.

\section{More advanced analysis}

\subsection{Time course with non-uniform sampling}
If the samples were collected with different time intervals and the user wants to
use the original time (instead of a vector of consecutive numbers),
the user may specify it via the T.Vect parameter in the \Rfunction{trendy} function.

For example, suppose for the example data, the first 30 samples were collected
every hour and the other 10 samples were collected every 5 hours. We may define
the time vector as:
<<eval=TRUE>>=
t.v <- c(1:30,seq(31,80,5))
names(t.v) <- colnames(TrendyExData)
print(t.v)
@

To run Trendy model using the empirical collecting time instead of sample ID (1-40):

<<eval=TRUE, warning=FALSE, fig.height=7, fig.width=7, fig.align='center', out.width='.8\\textwidth'>>=
res2 <- trendy(TrendyExData, T.Vect=t.v, Max.K=2, Cut.Diff=.05)
res.top2 <- toptrendy(res2)
res.trend2 <- trendheatmap(res.top2)
str(res.trend2)
@

To plot the first six genes that have up-regulated pattern at the beginning of the time course:
<<eval=TRUE>>=
par(mfrow=c(3,2))
plot1.new <- plotfeature(TrendyExData, T.Vect=t.v,
                        Feature.Names = names(res.trend2$firstup)[1:6], 
                        Seg.Data = res2)
@



\section{Extract genes with certain pattern}
Genes that have a peak along the time-course will have fitted trend somewhere as "up-down". Genes that are oscillating may have the fitted trend "up-down-up-down". To extract a list of such genes we can use the \Rfunction{extractpattern}:

<<eval=TRUE, warning=FALSE, fig.height=7, fig.width=7, fig.align='center', out.width='.8\\textwidth'>>=
# Genes that peak
pat1 <- extractpattern(res2, Pattern = c("up","down"))
head(pat1)
par(mfrow=c(3,2))
plotPat1 <- plotfeature(TrendyExData, T.Vect=t.v,
                      Feature.Names = pat1$Gene[1:6], 
                      Seg.Data = res2)

@

% # <<eval=TRUE, warning=FALSE, fig.height=7, fig.width=7, fig.align='center', out.width='.8\\textwidth'>>=
% # # Genes that oscillate
% # pat1 <- extractpattern(res3, Pattern = c("up","down","up","down"))
% # @

<<eval=TRUE, warning=FALSE, fig.height=3.5, fig.width=7, fig.align='center', out.width='.8\\textwidth'>>=
# Genes that peak after some time
pat3 <- extractpattern(res2, Pattern = c("up","down"), Delay = 25)
head(pat3)
par(mfrow=c(1,2))
plotPat3 <- plotfeature(TrendyExData, T.Vect=t.v,
                      Feature.Names = pat3$Gene, 
                      Seg.Data = res2)
@


\section{Additional options}
In the \Rfunction{trendy} function,
the thresholds $c_{num}$, $c_{diff}$ and  $c_{pval}$
can be specified via parameters
Min.Num.In.Seg, Cut.Diff and Pval.Cut, respectively.

\section{Trendy shiny app}

The Trendy shiny app requires an .RData object output from the \Rfunction{trendy} function, which can be obtained by setting \Rcode{Save.Object=TRUE}.

<<eval=TRUE, warning=FALSE>>=
res <- trendy(TrendyExData, Max.K=2, Save.Object = TRUE, File.Name="exampleObject")
@

Then in R run:
<<eval=FALSE, warning=FALSE>>=
library(shiny)
runGitHub('rhondabacher/Trendy')
@

\begin{figure}[H]
\centering
\includegraphics[width=1\textwidth]{Shiny_UploadData.png}
\caption{Upload shiny object}
\end{figure}


\begin{figure}[H]
\centering
\includegraphics[width=1\textwidth]{Shing_allGenePat.png}
\caption{Find all genes with a given pattern}
\end{figure}


\begin{figure}[H]
\centering
\includegraphics[width=1\textwidth]{Shiny_geneSearch.png}
\caption{Search genes individually}
\end{figure}

\newpage
\section{SessionInfo}
<<eval=TRUE, warning=FALSE>>=
print(sessionInfo())
@






\end{document}
